clustering/ejb/WFLY-16345_SFSB_externalization_to_remote_Infinispan_server.adoc

```diff
@@ -0,0 +1,91 @@
+= @Stateful Jakarta Enterprise Bean instance externalization to remote Infinispan cluster
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently, WildFly only supports the use of embedded Infinispan for persisting/replicating @Stateful Jakarta Enterprise Bean instances.
```

Comment:
Correct.

Reply:
@pferraro in a set of clustered WF nodes, what would be the difference(s) between using the new `hotrod-bean-management` and the existing `invalidation-cache`?

Reply:
Advantages of the RemoteCache-based managers (over their embedded invalidation-cache equivalents) include:

- Simpler load balancing requirements, i.e. no need for primary-owner based routing
- Low clustering overhead for EAP servers
  - No need for JGroups discovery, failure detection, merge/join handling, etc.
  - Faster to scale up/down EAP cluster

Disadvantages of the RemoteCache-based managers (over their embedded cache equivalents) include:

- Generally slower individual response times
  - Each request require extra network call (reads always go remote)
  - Near caching should be avoided unless EAP cluster size is small
- Inefficient expiration handling
  - Ensuring \@PreDestroy is invoked on a single cluster member is very expensive

Reply:
@pferraro IIUC the main use case for this feature is using standalone WildFly instances in a cloud environment lile OpenShift; at present, to cache SFSB in a remote Infinispan cluster, we would need to have  WildFly clustered which implies using a [`StafulSet`](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) instead of a simple [`Deployment`](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/); this feature would used in cross-site scenarios as well;
If that's correct perhaps we can add a simple test with our usual two nodes cluster connected to a remote Infinispan two nodes cluster and another test running on OpenShift;

Reply:
A more comprehensive example of a configuration we might use without this new feature is the following:
```
embed-server --server-config=standalone-ha.xml

/socket-binding-group=standard-sockets/remote-destination-outbound-socket-binding=remote-jdg-server1:add(host=10.0.145.66, port=11222)
/socket-binding-group=standard-sockets/remote-destination-outbound-socket-binding=remote-jdg-server2:add(host=10.0.145.120, port=11222)

# JDG Nodes
/subsystem=elytron/key-store=wildfly-truststore-containing-ca:add(path=wildfly-truststore.p12, relative-to=jboss.home.dir, credential-reference={clear-text=123PIPPOBAUDO}, type=PKCS12)
/subsystem=elytron/trust-manager=wildfly-trust-manager-containing-ca:add(key-store=wildfly-truststore-containing-ca, algorithm="SunX509")

/subsystem=elytron/client-ssl-context=CLIENT_SSL_CONTEXT:add(trust-manager=wildfly-trust-manager-containing-ca, protocols=["TLSv1.2"])
batch
/subsystem=infinispan/remote-cache-container=web-sessions:add(default-remote-cluster=jdg-server-cluster, statistics-enabled=true, properties={infinispan.client.hotrod.auth_username=admin, infinispan.client.hotrod.auth_password=pass.1234})
/subsystem=infinispan/remote-cache-container=web-sessions/remote-cluster=jdg-server-cluster:add(socket-bindings=[remote-jdg-server1,remote-jdg-server2])
run-batch
/subsystem=infinispan/remote-cache-container=web-sessions/component=security:write-attribute(name=ssl-context,value=CLIENT_SSL_CONTEXT)
if (outcome == success) of /subsystem=infinispan/remote-cache-container=web-sessions:read-attribute(name=marshaller)
/subsystem=infinispan/remote-cache-container=web-sessions:write-attribute(name=marshaller,value=PROTOSTREAM)
end-if
/subsystem=infinispan/cache-container=web/invalidation-cache=offload:add()
/subsystem=infinispan/cache-container=web/invalidation-cache=offload/store=hotrod:add(remote-cache-container=web-sessions, fetch-state=false, preload=false, passivation=false, purge=false, shared=true)
/subsystem=infinispan/cache-container=web/invalidation-cache=offload/component=transaction:add(mode=BATCH)

/subsystem=infinispan/cache-container=web:write-attribute(name=default-cache, value=offload)

/subsystem=infinispan/cache-container=ejb/invalidation-cache=offload:add()
/subsystem=infinispan/cache-container=ejb/invalidation-cache=offload/store=hotrod:add(remote-cache-container=web-sessions, fetch-state=false, preload=false, passivation=false, purge=false, shared=true)
/subsystem=infinispan/cache-container=ejb/invalidation-cache=offload/component=transaction:add(mode=BATCH)

/subsystem=infinispan/cache-container=ejb:write-attribute(name=default-cache, value=offload)
```